hudi-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadata.java

```diff
@@ -188,41 +196,51 @@ private synchronized void openFileSliceIfNeeded() throws IOException {
 
     // Load the schema
     Schema schema = HoodieAvroUtils.addMetadataFields(HoodieMetadataRecord.getClassSchema());
-    logRecordScanner = new HoodieMetadataMergedLogRecordScanner(metaClient.getFs(), metadataBasePath,
-            logFilePaths, schema, latestMetaInstantTimestamp, MAX_MEMORY_SIZE_IN_BYTES, BUFFER_SIZE,
+    HoodieMetadataMergedLogRecordScanner logRecordScanner = new HoodieMetadataMergedLogRecordScanner(metaClient.getFs(),
+            metadataBasePath, logFilePaths, schema, latestMetaInstantTimestamp, MAX_MEMORY_SIZE_IN_BYTES, BUFFER_SIZE,
             spillableMapDirectory, null);
 
     LOG.info("Opened metadata log files from " + logFilePaths + " at instant " + latestInstantTime
         + "(dataset instant=" + latestInstantTime + ", metadata instant=" + latestMetaInstantTimestamp + ")");
 
     metrics.ifPresent(metrics -> metrics.updateMetrics(HoodieMetadataMetrics.SCAN_STR, timer.endTimer()));
+
+    if (metadataConfig.enableReuse()) {
+      // cache for later reuse
+      cachedBaseFileReader = baseFileReader;
+      cachedLogRecordScanner = logRecordScanner;
+    }
+
+    return Pair.of(baseFileReader, logRecordScanner);
   }
 
-  private void closeIfNeeded() {
+  private void closeIfNeeded(Pair<HoodieFileReader, HoodieMetadataMergedLogRecordScanner> readers) {
     try {
       if (!metadataConfig.enableReuse()) {
-        close();
+        readers.getKey().close();
```

Comment:
I'm sure you have tried it. But not sure why can't we achieve this w/o two diff set of variables for both code paths? Can't we have just one set of reader variables. one which gets closed and reopened every time (if reuse is not enabled), or the same one getting re-used (if config is enabled)

Reply:
The reason is that we do not lock the entire function getRecordByKeyFromMetadata(). So if we have one set of the reader variables, one of the thread may be reading a key while the other thread may be calling a close() on the readers.

The getRecordByKeyFromMetadata() function does the following:
1. Get the correct readers (open new readers if reuse=false)
2. Read the key from the baseFileReader (reads key from HFile)
3. Bytes to HoodieRecordPayload conversion (bytes read in above step from HFile)
4. Read the key from the logRecordScanner (in-memory lookup)
5. Merge the two payloads to get the final value  
6. Close the readers (if reuse=false)

We should only lock during Step 2 as HFile KeyScanner is not thread-safe. Rest of the steps can be done by multiple threads in parallel for max performance.

Reply:
thnx for the detailed explanation Prasanth. will sync up w/ you offline. but still have some doubts/clarifications. 
wrt statement 'one thread could be reading a key, while another could be closing the readers".
Let's talk about 2 scenarios. 
a. reuse is not enabled. every thread is going to reinitialize the reader and close in the end and has a local copy of reader. And hence should not be an issue. 
b. reuse is enabled. close() is invoked only when the HoodieBackedMetadataTable itself will be closed. So, I thought no thread will invoke close() when some other thread is reading. 
May be I am missing something here. as I told, will sync up directly.